### PR TITLE
feat: add generateThemeStylesheet API, only output override styles

### PR DESCRIPTION
**Pull request by afsmith92**
This change adds the `generateThemeStylesheet` API for runtime theming. The currently available `applyTheme` API internally generates theme CSS and creates a `<style />` tag and inserts it. This `generateThemeStylesheet` API only generates the stylesheet content. `applyTheme` is an expensive call -- with this new API the styles can be generated server-side and cached (regardless of CSR/SSR).

As part of this contribution, I'm making an optimization to the generation of theme styles. Currently css is generated for every cloudscape design token (rather than just those that were overridden for theming) resulting in 900KB+ (90KB gzipped, ~20K lines) of mostly duplicate css. Additionally, the call time for `applyTheme` is relatively high. In our fargate container, this call takes ~800ms - 1000ms, and on my local machine takes ~150ms in the browser.

I've updated runtime theming to only generate css needed for the specified overrides. My assumption is that everyone using runtime theming would have base styles, but if we can't make this assumption I can instead expose this behavior as an opt-in flag.

For my particular use case (using ~45 design tokens), this results in override CSS of ~90KB (4KB gzipped, ~1500 lines). The call time is also reduced -- though not in proportion to the css size because iterating over all design tokens in the base theme is still required. In our fargate container, the call time is now ~100ms - 200ms, and is ~75ms on my local machine.

Will cut a PR to the `components` repo to expose `generateThemeStylesheet` once this is merged

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

### CODE REVIEW

**Comment by afsmith92** [comment]
Setting to draft -- found a behavioral difference with contexts

**Review by just-boris — COMMENTED** [review]
Still waiting for the regression tests to complete in my internal pipeline, I will let you know about the result

File: src/browser/index.ts [hunk 1]
```diff
@@ -6,32 +6,44 @@ import { getNonce, createStyleNode, appendStyleNode } from './dom';
 import { createMultiThemeCustomizer } from '../shared/declaration/customizer';
 import { getContexts, getThemeFromPreset } from '../shared/theme/validate';
 
-export interface ApplyThemeParams {
+export interface GenerateThemeStylesheetParams {
   override: Override;
   preset: ThemePreset;
   baseThemeId?: string;
   targetDocument?: Document;
```
[the remark this answers is not in the document]
Comment (afsmith92) [reply src/browser/index.ts:13]: no, will remove

File: src/browser/index.ts [hunk 2]
```diff
@@ -6,32 +6,44 @@ import { getNonce, createStyleNode, appendStyleNode } from './dom';
...
 }
 
-export interface ApplyThemeResult {
-  reset: () => void;
-}
-
-export function applyTheme(params: ApplyThemeParams): ApplyThemeResult {
-  const { override, preset, baseThemeId, targetDocument } = params;
+export function generateThemeStylesheet(params: GenerateThemeStylesheetParams): string {
```
[the remark this answers is not in the document]
Comment (afsmith92) [reply src/browser/index.ts:16]: My intention was to expose `generateThemeStylesheet` from the `theming-runtime` package. Looks like if I move it to `src/build` it would be instead exposed via `theming-build`. IMO that doesn't fit with the use case for this function -- but are you saying regardless of the package names the packages are really `theming-browser` and `theming-node`? FWIW, in my testing simply importing `generateThemeStylesheet` on the server doesn't break anything because the DOM API dependencies would only become an issue if `applyTheme` were actually called server side. (But perhaps you're concerned with future potential of needing to use side effects in this module?) Also, while it makes the most sense to call `generateThemeStylesheet` server side, there's technically nothing stopping anyone from calling it client side

Either way, happy to move it once you confirm

**Comment by afsmith92** [comment]
Also, a question about theming behavior for contexts. I see that the current behavior of the API is -- when resolving contexts which refer to base tokens -- to use the default theme tokens rather than the override theme tokens. Is there a reason this is the case?

For example, with the `top-navigation` context, it overrides a couple tokens and then [also overrides](https://github.com/cloudscape-design/components/blob/34b4ff4fa8e0511a67dafc2cdd68d7553ddb926d/style-dictionary/visual-refresh/contexts/top-navigation.ts#L13) all of the color tokens from the base theme to their dark mode values.

If I were to call
```javascript
applyTheme({
  theme: {
    tokens: {
      colorBackgroundExample: {
        light: 'green',
        dark: 'orange'
      }
    }
  },
  'top-navigation': {} // not passing in any context overrides
})
```

This would not use `orange` for `colorBackgroundExample` in the `top-navigation` context but would instead use the default VR value for `colorBackgroundExample.dark`

Similarly, if set `colorBackgroundStatusError` in `tokens`, it won't take effect for the `Alert` component unless I specify that token in the `alert` context even though the alert context [uses the default/root](https://github.com/cloudscape-design/components/blob/34b4ff4fa8e0511a67dafc2cdd68d7553ddb926d/style-dictionary/visual-refresh/contexts/alert.ts#L57) VR value for `colorBackgroundStatusError`

So basically it seems as though I am required to define every context instead of (at least having a chance at) inheriting behavior that cloudscape uses when referring to the base theme. 

With this PR I haven't modified this behavior, but just wanted to see what you thought about this.

File: src/browser/index.ts (hunk 2 above)
[the remark this answers is not in the document]
Comment (just-boris) [reply src/browser/index.ts:15]: I see, so you are going to call this function in isomorphic code (something like `app/layout.js` in next). You do not want to pull build time dependencies in that closure, that's true.

I guess, we will end up exposing it in both, but let's keep your current option with `src/browser/index.ts`

[the remark this answers is not in the document]
Comment (just-boris) [reply src/browser/index.ts:15]: P.S. While we are at this thread, can we also have more tests on `generateThemeStylesheet` standalone?

**Comment by afsmith92** [comment]
> If you see a potential to reduce output CSS by this refactoring, we can discuss this, but it will require changes our existing code in a bunch of places.

Oh I was actually just asking because I was initially confused about how to use the API -- that wasn't about CSS size. I had thought that any values I put into `tokens` would be automatically inherited in `contexts` -- so I was confused when setting `colorBackgroundStatusError` in `tokens` didn't result in any changes in the `Alert` component.

From what I can tell, changing that behavior (while it might be a bit more intuitive IMO) would not reduce the CSS size at all. In the code comment block on the `resolveContext` function I described an optimization that it looks like cloudscape does: when a theme context token has the same assignment as the base token, cloudscape only generates css for the base token -- that's why I had to add that baselining logic in `resolveContext`.

File: src/shared/theme/resolve.ts [hunk 3]
```diff
@@ -69,34 +90,79 @@ function resolveToken(theme: Theme, token: string, path: Array<string>, state?:
...
+        ? acc
+        : {
+            ...acc,
+            [key]: baseContext.tokens[key] ?? theme.tokens[key] ?? baseTheme.tokens[key],
+          };
+    }, {}),
     ...context.tokens,
   };
-  return resolveTheme(tmp);
+
```
Comment (afsmith92) [inline src/shared/theme/resolve.ts:147]: I debated adding another optimization here, but I questioned whether I should rely on the css optimization that cloudscape does (described in the `resolveContext` code comment) so I ultimately decided to leave it out. But in my testing with the ~45 tokens I was using, it reduced the css size from ~120KB to ~70KB. The gzip value remained the same at 4KB, so this optimization would only mean less css to cache server side but probably wouldn't change TTI, etc. much

```typescript
/**
 * Relying on the "override theme tokens incorrectly taking precedence over base theme
 * context" behavior described above when it is advantageous
 */
Object.keys(context.tokens).forEach((key) => {
  const isDuplicateAssignment =
    areAssignmentsEqual(context.tokens[key], theme.tokens[key]) &&
    (!(key in baseContext.tokens) || areAssignmentsEqual(baseContext.tokens[key], baseTheme.tokens[key]));

  if (isDuplicateAssignment) {
    delete tmp.tokens[key];
  }
});
```

**Review by just-boris — APPROVED** [review]

File: src/shared/theme/resolve.ts (hunk 3 above)
Comment (just-boris) [reply src/shared/theme/resolve.ts:147]: We came up with the same outcome last time we looked into that. We can possibly look into refactoring this all together, to not produce duplicate values from the start, but "inflate then clean" pattern is more prone to errors, so let's keep it this way.

**Comment by just-boris** [comment]
Available in [v1.0.56](https://github.com/cloudscape-design/theming-core/releases/tag/1.0.56)

**Comment by afsmith92** [comment]
Thanks @just-boris -- pull request on components repo is open
